### ISSUE TITLE
[JENKINS-22909] git plugin does not support username/password over ssh

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/git-client-plugin#1439]
<p>We are trying to connect to a Git repository via SSH (not Github) using username/password (we're not using keys) and we're getting the following error:</p>

<p>Failed to connect to repository : Command "git ls-remote -h ssh://account@​repository/path/toApplication HEAD" returned status code 128:<br/>
stdout: <br/>
stderr: Permission denied, please try again. <br/>
Permission denied, please try again. <br/>
Permission denied (publickey,gssapi-keyex,gssapi-with-mic,password). <br/>
fatal: Could not read from remote repository.</p>

<p>Please make sure you have the correct access rights<br/>
and the repository exists.<br/>
------------------</p>

<p>If we use the anonymous access or change the access type from SSH to HTTPS, it works fine, but we should be able to use the SSH with user/password without having to add the public/private keys.</p>

<p>Jenkins ver. 1.532.1<br/>
Git Client Plugin 1.6.4    <br/>
Git Plugin 2.0.4 <br/>
•How are you running Jenkins? <br/>
	We are running  Jenkins as a service in RHEL<br/>
•Have you specified additional parameters for the Java VM (Heapspace etc.)?<br/>
	No<br/>
•Did you just install the deb or rpm?<br/>
	We used the RPM jenkins-1.532.1-1.1.noarch.rpm<br/>
•With which Java VM (Oracle, IBM etc.)?<br/>
	java version "1.7.0_45"<br/>
	Java(TM) SE Runtime Environment (build 1.7.0_45-b18)<br/>
	Java HotSpot(TM) 64-Bit Server VM (build 24.45-b08, mixed mode)<br/>
•On which operating system? 32- or 64-bit?<br/>
	Linux version 2.6.32-358.23.2.el6.x86_64 (mockbuild@​ca-build44.us.oracle.com) (gcc version 4.4.7 20120313 (Red Hat 4.4.7-3) (GCC) )</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=agalaviz">agalaviz</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-22909" target="_blank">git plugin does not support username/password over ssh</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=agalaviz">agalaviz</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: git-client-plugin
<li><b>resolution</b>: Won't Do
<li><b>resolved</b>: 2022-11-23T05:27:42+00:00
<li><b>votes</b>: 4
<li><b>watchers</b>: 13
<li><b>imported</b>: 20251211-071809
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>We are trying to connect to a Git repository via SSH (not Github) using username/password (we're not using keys) and we're getting the following error:</p>

<p>Failed to connect to repository : Command "git ls-remote -h ssh://account@repository/path/toApplication HEAD" returned status code 128:
stdout: 
stderr: Permission denied, please try again. 
Permission denied, please try again. 
Permission denied (publickey,gssapi-keyex,gssapi-with-mic,password). 
fatal: Could not read from remote repository.</p>

<p>Please make sure you have the correct access rights
and the repository exists.
------------------</p>

<p>If we use the anonymous access or change the access type from SSH to HTTPS, it works fine, but we should be able to use the SSH with user/password without having to add the public/private keys.</p>

<p>Jenkins ver. 1.532.1
Git Client Plugin 1.6.4    
Git Plugin 2.0.4 
•How are you running Jenkins? 
	We are running  Jenkins as a service in RHEL
•Have you specified additional parameters for the Java VM (Heapspace etc.)?
	No
•Did you just install the deb or rpm?
	We used the RPM jenkins-1.532.1-1.1.noarch.rpm
•With which Java VM (Oracle, IBM etc.)?
	java version "1.7.0_45"
	Java(TM) SE Runtime Environment (build 1.7.0_45-b18)
	Java HotSpot(TM) 64-Bit Server VM (build 24.45-b08, mixed mode)
•On which operating system? 32- or 64-bit?
	Linux version 2.6.32-358.23.2.el6.x86_64 (mockbuild@ca-build44.us.oracle.com) (gcc version 4.4.7 20120313 (Red Hat 4.4.7-3) (GCC) )</p></pre>
</details>
</details>
<details><summary><i>1 attachment</i></summary>

- [Jenkins System Properties.docx](https://issues.jenkins.io/secure/attachment/25776/Jenkins%20System%20Properties.docx)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-22909?focusedId=199876&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-199876">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Embedding the user name and password in the URL directly will expose your user name and password in the job definition and may expose it in the job log.  If you'll enter a credential with that same user name and password, you can then reference that credential without having the user name or the password visible in the job definition or in the console log from the build.</p>

<p>I agree that the plugin should also accept the user name and password embedded in the URL, but it does not currently.  The work around of entering the user name and password as a credential is workable, and is more secure.</p>
</pre>
</details>

</ul>
</details>

<p>Embedding the user name and password in the URL directly will expose your user name and password in the job definition and may expose it in the job log.  If you'll enter a credential with that same user name and password, you can then reference that credential without having the user name or the password visible in the job definition or in the console log from the build.</p>

<p>I agree that the plugin should also accept the user name and password embedded in the URL, but it does not currently.  The work around of entering the user name and password as a credential is workable, and is more secure.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-22909?focusedId=218121&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-218121">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The more I look at this report, the less confident I am that I understand what is requested.</p>

<p>I'm able to clone with a username / password credential without using public / private keys with the following steps:</p>

<ol>
	<li>Configure a global username / password credential (through "Manage Credentials") with appropriate user name and password (no public key, no private key)</li>
	<li>Create a new freestyle job - <a href="https://issues.jenkins.io/browse/JENKINS-22909" title="git plugin does not support username/password over ssh" class="issue-link" data-issue-key="JENKINS-22909"><del>JENKINS-22909</del></a>-username-password-credentials</li>
	<li>Use git as the SCM for the job</li>
	<li>Use an ssh URL for the repository. I used ssh://mwaite@mark-pc1.markwaite.net/var/lib/git/mwaite/bin.git</li>
	<li>Use the previously defined username / password credential as the credential for that ssh URL</li>
	<li>Run the job, confirm it succeeds</li>
</ol>


<p>I suspect that my clone succeeds through an accident in that case.  The accident is that my jenkins user is likely allowed to access that repository as user mwaite through ssh by default.  When I run in a "clean" environment, username/password authentication fails for an ssh protocol URL unless I also use a private key.</p>

<p>Are there more steps you're taking to show the problem?</p>

<p>Have you used a more recent git client plugin and a more recent git plugin?</p>
</pre>
</details>

</ul>
</details>

<p>The more I look at this report, the less confident I am that I understand what is requested.</p>

<p>I'm able to clone with a username / password credential without using public / private keys with the following steps:</p>

<ol>
	<li>Configure a global username / password credential (through "Manage Credentials") with appropriate user name and password (no public key, no private key)</li>
	<li>Create a new freestyle job - <a href="https://issue-redirect.jenkins.io/issue/22909" title="git plugin does not support username/password over ssh" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/22909">JENKINS-22909</a>"><del>JENKINS-22909</del></a>-username-password-credentials</li>
	<li>Use git as the SCM for the job</li>
	<li>Use an ssh URL for the repository. I used ssh://mwaite@​mark-pc1.markwaite.net/var/lib/git/mwaite/bin.git</li>
	<li>Use the previously defined username / password credential as the credential for that ssh URL</li>
	<li>Run the job, confirm it succeeds</li>
</ol>


<p>I suspect that my clone succeeds through an accident in that case.  The accident is that my jenkins user is likely allowed to access that repository as user mwaite through ssh by default.  When I run in a "clean" environment, username/password authentication fails for an ssh protocol URL unless I also use a private key.</p>

<p>Are there more steps you're taking to show the problem?</p>

<p>Have you used a more recent git client plugin and a more recent git plugin?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>sabareesh_ss:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-22909?focusedId=223916&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-223916">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I am also following the same steps. I tried all my best. My build failed and ended up with </p>

<p>"Failed to connect to repository : Command "C:\Program Files\Git\bin\git.exe -c core.askpass=true ls-remote -h ssh://git@localhost:7999/jen/jensh_empty.git HEAD" returned status code 128:
stdout: 
stderr: Permission denied (publickey). 
fatal: Could not read from remote repository.</p>

<p>Please make sure you have the correct access rights
and the repository exists."</p>

<p>Please help.</p>

<p>How can I fix this issue?</p>
</pre>
</details>

</ul>
</details>

<p>I am also following the same steps. I tried all my best. My build failed and ended up with </p>

<p>"Failed to connect to repository : Command "C:\Program Files\Git\bin\git.exe -c core.askpass=true ls-remote -h ssh://git@​localhost:7999/jen/jensh_empty.git HEAD" returned status code 128:<br/>
stdout: <br/>
stderr: Permission denied (publickey). <br/>
fatal: Could not read from remote repository.</p>

<p>Please make sure you have the correct access rights<br/>
and the repository exists."</p>

<p>Please help.</p>

<p>How can I fix this issue?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>sabareesh_ss:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-22909?focusedId=223917&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-223917">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The issue is reproduced n number of times.</p>

</pre>
</details>

</ul>
</details>

<p>The issue is reproduced n number of times.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>tkp75:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-22909?focusedId=246280&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-246280">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Still able to reproduce this with the latest available versions:
Jenkins ver. 1.644
Credentials Plugin 1.24
Git plugin 2.4.1
Git client plugin 1.19.1</p>

<p>I used RHEL 7.2 (x86_64) with openssh:
openssh.x86_64                                       6.6.1p1-23.el7_2
openssl.x86_64                                       1:1.0.1e-51.el7_2.2</p>
</pre>
</details>

</ul>
</details>

<p>Still able to reproduce this with the latest available versions:<br/>
Jenkins ver. 1.644<br/>
Credentials Plugin 1.24<br/>
Git plugin 2.4.1<br/>
Git client plugin 1.19.1</p>

<p>I used RHEL 7.2 (x86_64) with openssh:<br/>
openssh.x86_64       6.6.1p1-23.el7_2<br/>
openssl.x86_64       1:1.0.1e-51.el7_2.2</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-22909?focusedId=246518&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-246518">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>In all my recent tests of credentials, I've never seen a case where an ssh URL was able to authenticate with a user name and password.  When using an ssh URL, the plugin seems to expect that you are using a private key for authentication, rather than user name and password.</p>
</pre>
</details>

</ul>
</details>

<p>In all my recent tests of credentials, I've never seen a case where an ssh URL was able to authenticate with a user name and password.  When using an ssh URL, the plugin seems to expect that you are using a private key for authentication, rather than user name and password.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jimklo:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-22909?focusedId=246538&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-246538">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'm not sure if this is related, however it sound's very similar: <a href="https://issues.jenkins-ci.org/browse/JENKINS-24368?focusedCommentId=246537&amp;page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-246537" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-24368?focusedCommentId=246537&amp;page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-246537</a></p>
</pre>
</details>

</ul>
</details>

<p>I'm not sure if this is related, however it sound's very similar: <a href="https://issues.jenkins-ci.org/browse/JENKINS-24368?focusedCommentId=246537&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-246537" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-24368?focusedCommentId=246537&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-246537</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-22909?focusedId=246606&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-246606">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jimklo" class="user-hover" rel="jimklo">jimklo</a> I don't think they are related, since I believe they use different paths through the code for ssh based authentication and https based authentication.  <a href="https://issues.jenkins.io/browse/JENKINS-24368" title="Authentication failed for https://username@bitbucket.org/project/repository.git" class="issue-link" data-issue-key="JENKINS-24368"><del>JENKINS-24368</del></a> is specific to https. (and seems to be specific to bitbucket)  I can connect to a bitbucket repo with ssh, just not with https in certain configurations.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jimklo" class="user-hover" rel="jimklo">jimklo</a> I don't think they are related, since I believe they use different paths through the code for ssh based authentication and https based authentication.  <a href="https://issue-redirect.jenkins.io/issue/24368" title="Authentication failed for https://username@​bitbucket.org/project/repository.git" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/24368">JENKINS-24368</a>"><del>JENKINS-24368</del></a> is specific to https. (and seems to be specific to bitbucket)  I can connect to a bitbucket repo with ssh, just not with https in certain configurations.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jaswo:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-22909?focusedId=249370&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-249370">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I do not want to sound pushy, but this issue was created since May 2014 and its now March 2016. Can someone that is responsible for resolving this bug give an indication as to when it will be resolved? Thanks!</p>
</pre>
</details>

</ul>
</details>

<p>I do not want to sound pushy, but this issue was created since May 2014 and its now March 2016. Can someone that is responsible for resolving this bug give an indication as to when it will be resolved? Thanks!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-22909?focusedId=249379&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-249379">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jaswo" class="user-hover" rel="jaswo">jaswo</a> I'm probably the only one that is close to "responsible for resolving this bug", so I'll give you my estimate.  I don't know that it will be resolved any time within the next few years.  A private key works quite well with ssh protocol, so I perceive there is a relatively simple work around to allow access to ssh protected git repositories.</p>

<p>I'm currently focused on other areas, like adding authenticated submodule access, and like processing the ~50 pull requests pending on the git plugin, and the ~5 pull requests pending on the git client plugin.  </p>

<p>My first goal is to not regress functionality for the existing 70000+ installations of the plugin.  I'd love to have additional help from others evaluating the pull requests, and submitting proposed improvements (fixes or features) with tests to support those additions.  I'd also love to have additional automated tests for the existing features.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jaswo" class="user-hover" rel="jaswo">jaswo</a> I'm probably the only one that is close to "responsible for resolving this bug", so I'll give you my estimate.  I don't know that it will be resolved any time within the next few years.  A private key works quite well with ssh protocol, so I perceive there is a relatively simple work around to allow access to ssh protected git repositories.</p>

<p>I'm currently focused on other areas, like adding authenticated submodule access, and like processing the ~50 pull requests pending on the git plugin, and the ~5 pull requests pending on the git client plugin.  </p>

<p>My first goal is to not regress functionality for the existing 70000+ installations of the plugin.  I'd love to have additional help from others evaluating the pull requests, and submitting proposed improvements (fixes or features) with tests to support those additions.  I'd also love to have additional automated tests for the existing features.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>battika:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-22909?focusedId=274235&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-274235">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I have just encountered the same issue but happens only on Windows Slave if using  SSH username thus credentialsID with the git plugin. On Linux Slave works, and as said it also works with ssh keys. So my workaround was to go with the ssh keys.</p>
</pre>
</details>

</ul>
</details>

<p>I have just encountered the same issue but happens only on Windows Slave if using  SSH username thus credentialsID with the git plugin. On Linux Slave works, and as said it also works with ssh keys. So my workaround was to go with the ssh keys.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-22909?focusedId=432638&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-432638">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Won't implement this enhancement request.</p>
</pre>
</details>

</ul>
</details>

<p>Won't implement this enhancement request.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `duplicates` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/28165">JENKINS-28165</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/28165">JENKINS-28165</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/22855">JENKINS-22855</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/24368">JENKINS-24368</a>]</i>
